bpf: add LB ipip health check datapath

borkmann opened the pull request:
(See commit msgs.)

borkmann committed:
bpf: rename current set of bind helpers to post_bind

Denote that this happens after bind assigned an IP/port where we reject
this ... given we're adding pre-bind code, document the code as such by
renaming.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add get_socket_opt helper definition

Fyi, we remap it here to avoid getsockopt(2) clash from bpf/api include
of regular user applications like cilium-map-migrate.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add stand-alone load-balancer ipip health check datapath

Add the possibility to health probe backends through ipip encap so they
reach endpoints in the same way as regular datapath traffic.

There are various options on how this could be achieved. One possibility
we considered was to use SRIOV where we'd spawn a VF and inject traffic
into the regular XDP datapath at the external facing device by rewriting
the dst MAC to the one of the PF at VF tc egress time. This traffic then
reaches the XDP prog and can be pushed out the same devices for leaving
the node. This option has one downside though, that is, it's limited to
bare metal deployment scenarios (it's not possible to spawn a VF from VF
when in off-prem / cloud) and the HW needs to support SRIOV. Performance
wise, the use of XDP_TX has no downgrade under SRIOV, but XDP_REDIRECT
has. In Cilium case we only use XDP_TX today, but in future we potentially
could also extend for multi-dev XDP_REDIRECT case.

This led us to a more flexible tc BPF based approach which does not have
the latter limitations. We do rely on ipip / ip6ip6 tunnel driver in
collect metadata mode though, meaning, this adds cilium_ipip4, cilium_ipip6
and cilium_sit devices. From kernel side, ipip driver adds the three netdevs
as a pair. One more design consideration is that the health check probes
are decoupled from the BPF service map, meaning, the service map is not
consulted on purpose in the datapath - this has the advantage that backends
can be independently checked from the service map without having to have a
'staging phase' for the service map to push them live after initial probe.
Additionally, the service map doesn't need a more complex synchronization
scheme with the health check logic to prevent backend position reordering
during a health probe run (for telling which of the backend failed when
they would be picked based on socket mark, for example).

That said, if the health checker residing on the LB node would initiate a
connection to svcIP:port, we still need a way to pass-in a backend IP that
we intend to test for availability, so that the ipip outer header can then
be populated as [L2][lbIP -> backendIP][lbIP -> svcIP][L4]. Typical client
applications allocate a socket and then call connect(2) or sendmsg(2) for
the probe. The connect(2) call will auto-bind the socket in the kernel
for obtaining a source port. This could be seen as equivalent of calling
bind(0.0.0.0:0) on the client before calling connect(2) or sendmsg(2). We
can piggy-back on this, mark a socket for getting special treatment for
the health check datapath and then do i) bind(backendIP:port) with ii)
subsequent connect(serviceIP:port). The bind()ing to a backend approach
allows for BPF at the pre-bind hook to record a IPv4 or IPv6 backend address
to be recorded in an LRU map based on the socket cookie where later on a
tc egress BPF program can retrieve that information from the skb (again,
via socket cookie) to populate the collect metadata tunnel key. The pre-bind
hook in the kernel allows to override the struct sockaddr, which allows to
clear the IP/port in order to let the kernel perform auto-bind. The socket
itself is locked during this time. This allows to define a simple API for
backend health probes.

Note that we mark skbs with MARK_MAGIC_IPIP_DONE once we set the tunnel key.
This is to prevent potential looping: The packet goes out the single phys
dev (given XDP it will be the one of the LB). There, on egress we lookup the
map to see if the skb is eligible for the ipip encap. If so, we populate
tunnel meta data and push it to the tunnel dev. From there it will do the
encap, then send it out the usual path via stack. We'll end up on the same
phys dev again and need to bypass the route via tunnel dev in this case.

Arguments used for testing:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true \
       --datapath-mode=lb-only --bpf-lb-algorithm=maglev        \
       --bpf-lb-maglev-table-size=2039 --bpf-lb-mode=dsr        \
       --bpf-lb-acceleration=native --devices=enp2s0np0         \
       --bpf-lb-dsr-dispatch=ipip

Example health probe client in C has been added outside of the main repo:

  https://github.com/cilium/health-examples

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add upstream linux/socket.h uapi header

This is needed for being able to use bpf_getsockopt(), bpf_setsockopt()
helpers out of sock_addr programs.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: remove sock_addr ctx workaround causing different alignment

Fix up sock_addr ctx workaround to include __bpf_md_ptr(). Including the
latter changes the natural struct alignment of struct bpf_sock_addr which
can cause verifier issues due to different LLVM code gen such as ctx access
optimizations. With that removed, the header is then on par with upstream.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: resync uapi header

Small resync of the uapi header which fixes up bpf_ prefixed helper
description before it got frozen by the final release [0] and a number
of new extensions and helpers.

  [0] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=b7906b70a2337e445b8dca3ce7ba8976b6ebd07d

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: exact word match on bpf_clear_cgroups for hook

Change the ID retrival of bpf_clear_cgroups() to be an exact match on
the hook to avoid getting the wrong ID. Example is bind{4,6} hook versus
post_bind{4,6} given the former is a substring of the latter.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

jrfastab commented on bpf/bpf_sock.c:
```diff
@@ -464,6 +477,44 @@ int sock4_post_bind(struct bpf_sock *ctx)
+
+	ret = map_update_elem(&LB4_HEALTH_MAP, &key, &val, 0);
+	if (!ret)
+		sock4_auto_bind(ctx);
```
Should we throw a bind error if ret != 0 here? Otherwise things are going to go a bit sideways later. If the pre_bind fails then the cilium health check tool can give some meaningful error.

borkmann replied:
good idea, and it indeed will get propagated, done

jrfastab commented on bpf/bpf_sock.c:
```diff
@@ -464,6 +477,44 @@ int sock4_post_bind(struct bpf_sock *ctx)
+	    !ctx_in_hostns(ctx, NULL))
+		return SYS_PROCEED;
+	if (sock_is_health_check(ctx))
+		__sock4_pre_bind(ctx, ctx);
```
check the error here and return SYS_REJECT?

jrfastab commented on bpf/bpf_sock.c:
```diff
@@ -772,6 +823,80 @@ int sock6_post_bind(struct bpf_sock *ctx)
+	    !ctx_in_hostns(ctx, NULL))
+		return SYS_PROCEED;
+	if (sock_is_health_check(ctx))
+		__sock6_pre_bind(ctx);
```
same here sys_reject on error?

jrfastab commented on bpf/lib/nodeport.h:
```diff
@@ -1678,6 +1678,97 @@ int tail_handle_nat_fwd_ipv4(struct __ctx_buff *ctx)
+		key = get_socket_cookie(ctx);
+		val = map_lookup_elem(&LB4_HEALTH_MAP, &key);
+		if (!val)
+			return CTX_ACT_OK;
```
return an error? This is bogus right, it seems to indicate key is broken or LRU is overrun or something? Might be nice to get an event this happened from caller of lb_handle_health.

borkmann replied:
It's needed, it would just mean that the given skb is not subject to health traffic, e.g. ssh connection to the LB or such. The call-site of this is from bpf_host on tc egress of the phys dev, so it would see all traffic leaving that box and health-traffic is a potentially small subset.

jrfastab commented on bpf/lib/nodeport.h:
```diff
@@ -1678,6 +1678,97 @@ int tail_handle_nat_fwd_ipv4(struct __ctx_buff *ctx)
+		key = get_socket_cookie(ctx);
+		val = map_lookup_elem(&LB6_HEALTH_MAP, &key);
+		if (!val)
+			return CTX_ACT_OK;
```
same as above

borkmann replied:
same as https://github.com/cilium/cilium/pull/14610#discussion_r560001697

jrfastab commented on bpf/lib/nodeport.h:
```diff
@@ -1678,6 +1678,97 @@ int tail_handle_nat_fwd_ipv4(struct __ctx_buff *ctx)
+		ret = health_encap_v6(ctx, &val->peer.address, 0);
+		if (ret != 0)
+			return ret;
+		ctx->mark |= MARK_MAGIC_IPIP_DONE;
```
This mark is just for tracing?

borkmann replied:
No, this is to prevent potential looping. The packet would go out the single phys dev (given XDP it will be the one of the LB). There on egress we lookup the map to see if the skb is eligible for the ipip encap. If so, we populate tunnel meta data and push it to the tunnel dev. From there it will do the encap, then then send it out the usual path via stack. We'll end up on the same phys dev again and need to bypass in this case.

borkmann replied:
Updated the commit msg to add this as information there.

jrfastab commented on bpf/lib/common.h:
```diff
@@ -445,6 +446,16 @@ enum {
 #define MARK_MAGIC_SNAT_DONE		0x1500
 
+/* MARK_MAGIC_IPIP_DONE can overlap with MARK_MAGIC_SNAT_DONE with both being
+ * mutual exclusive.
```
This is used to ...

borkmann replied:
So the IPIP dispatch requires DSR and for this we do not enable NAT code hence mutually exclusive.

jrfastab replied:
Sorry comment was short, I was just hinting that maybe the comment could have a bit extra detail about how the define is used.

borkmann replied:
I've updated the commit message with more details, but I could also extend the comment slightly. Can do after CI went through before merging.

jrfastab commented on bpf/lib/common.h:
```diff
@@ -445,6 +446,16 @@ enum {
+#define MARK_MAGIC_IPIP_DONE		MARK_MAGIC_SNAT_DONE
+
+/* MARK_MAGIC_HEALTH can overlap with MARK_MAGIC_DECRYPT with both being
+ * mutual exclusive.
```
Not going to NAK based on this but IMO we need to unwind this decrypt mutual exclusive problem at some point.

jrfastab reviewed (approved):
Nice looks good and clean to me. Couple comments about error handling so we fail a bit harder on errors. And general whining about what to do with encryption/node_port mutual exclusion problem.

Thanks for breaking the patch series up and detailed commit message made it easy to review.

brb commented on pkg/datapath/linux/config/config.go:
```diff
@@ -240,16 +240,27 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	}
 
 	if option.Config.EnableNodePort {
+		healthCheck := option.Config.DatapathMode == datapathOption.DatapathModeLBOnly &&
```
Shall we move into a helper function? We have an identical check in the ` pkg/datapath/loader/base.go`.

borkmann replied:
I just moved this into a config bool

brb commented on bpf/lib/lb.h:
```diff
@@ -59,6 +59,16 @@ struct bpf_elf_map __section_maps LB6_SRC_RANGE_MAP = {
 
+#ifdef ENABLE_HEALTH_CHECK
+struct bpf_elf_map __section_maps LB6_HEALTH_MAP = {
+	.type		= BPF_MAP_TYPE_LRU_HASH,
```
We should define both maps in the cilium-agent too (see https://github.com/cilium/cilium/issues/9276). Can be done as a follow up.

borkmann replied:
Defined names in agent and also added the cleanup just now.

brb commented on bpf/lib/lb.h:
```diff
@@ -131,6 +141,16 @@ struct bpf_elf_map __section_maps LB4_SRC_RANGE_MAP = {
+#ifdef ENABLE_HEALTH_CHECK
+struct bpf_elf_map __section_maps LB4_HEALTH_MAP = {
+	.type		= BPF_MAP_TYPE_LRU_HASH,
+	.size_key	= sizeof(__u64),
```
Maybe we could add a small comment saying that the key is a socket cookie?

borkmann replied:
I can clean it up and do a typedef, so all the locations are more clear. Same for netns cookie.

brb commented on bpf/lib/common.h:
```diff
@@ -445,6 +446,16 @@ enum {
+/* MARK_MAGIC_IPIP_DONE can overlap with MARK_MAGIC_SNAT_DONE with both being
+ * mutual exclusive.
+ */
+#define MARK_MAGIC_IPIP_DONE		MARK_MAGIC_SNAT_DONE
```
This is going to be used only for health check traffic. Maybe we should rename the macro to reflect its usage better? E.g. `MARK_MAGIC_LB_HEALTHCHECK_IPIP_DONE` or similar?

borkmann replied:
Bit long for my taste, but I could do a `MARK_MAGIC_LB_HEALTH_IPIP_DONE`.

brb commented on bpf/init.sh:
```diff
@@ -438,12 +438,41 @@ move_local_rules
-	echo "#define ENCAP_VXLAN 1" >> $RUNDIR/globals/node_config.h
-elif [ "$MODE" = "geneve" ]; then
-	echo "#define ENCAP_GENEVE 1" >> $RUNDIR/globals/node_config.h
+if [ "$MODE" = "ipip" ]; then
```
Should we rename the mode to `lb-ipip`? The current name might add some confusion (users might think that Cilium supports IPIP for endpoints connectivity).

borkmann replied:
Well, we'd need IPIP soon enough anyway for worker nodes, so this can be reused. But either way, this define is not used anywhere at this point.

brb commented on bpf/init.sh:
```diff
@@ -438,12 +438,41 @@ move_local_rules
+	if [ "$IP4_HOST" != "<nil>" ]; then
+		ENCAP_DEV="cilium_ipip4"
+		ip link show $ENCAP_DEV || {
+			ip link add name tunl0 type ipip external || true
```
Q: why do we set the iface name in the second step?

borkmann replied:
It's a bit ugly from kernel side. The issue is if you directly specify `cilium_ipip4` here, then the kernel will create a `tunl0` and `cilium_ipip4` device. The `tunl0` given it's the default when driver gets loaded, and one cannot remove this link. But it is definitely possible to rename which is done here w/o creating useless additional devices. I can add a comment on why we do this.

brb commented on bpf/init.sh:
```diff
@@ -438,12 +438,41 @@ move_local_rules
+	if [ "$IP6_HOST" != "<nil>" ]; then
+		ENCAP_DEV="cilium_ipip6"
+		ip link show $ENCAP_DEV || {
+			ip link add name ip6tnl0 type ip6tnl external || true
```
What kernel config does this require?
```
> sudo ip link add name ip6tnl0 type ip6tnl external
Error: Unknown device type.

> zgrep -i ipip /proc/config.gz
CONFIG_NET_IPIP=m
```

borkmann replied:
It's via `CONFIG_IPV6_TUNNEL`.

brb reviewed (commented):
Nice! A few nits.

borkmann committed:
bpf: add __net_cookie type

Make the code more self-descriptive and add a type mapper from __u64
to __net_cookie.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add __sock_cookie

Same as with __net_cookie, make the code more self-descriptive and
add a type mapper from __u64 to __sock_cookie so it is semantically
more clear than __u64.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
